src/main/java/com/aws/greengrass/secretmanager/SecretManager.java

```diff
@@ -154,23 +159,24 @@ public void reloadCache() throws SecretManagerException {
     * arn1:l1 -> secret3
     * arn1:l2 -> secret4
     */
-    private void loadCache(AWSSecretResponse awsSecretResponse)
-            throws SecretManagerException {
-        GetSecretValueResponse decryptedResponse = null;
-        try {
-            decryptedResponse = localStoreMap.decrypt(awsSecretResponse);
-        } catch (SecretCryptoException e) {
-            // This should never happen ideally
-            logger.atError().kv("secret",
-                    awsSecretResponse.getArn()).cause(e).log("Unable to decrypt secret, skip loading in cache");
-            throw new SecretManagerException("Cannot load secret from disk", e);
-        }
-        String secretArn = decryptedResponse.arn();
-        cache.put(secretArn, decryptedResponse);
-        cache.put(secretArn + decryptedResponse.versionId(), decryptedResponse);
-        // load all labels attached with this version of secret
-        for (String label : decryptedResponse.versionStages()) {
-            cache.put(secretArn + label, decryptedResponse);
+    private void loadCache(AWSSecretResponse awsSecretResponse) {
+        try (LockScope ls = LockScope.lock(cacheLock)) {
+            GetSecretValueResponse decryptedResponse = null;
+            try {
+                decryptedResponse = localStoreMap.decrypt(awsSecretResponse);
+            } catch (SecretCryptoException e) {
+                // This should never happen ideally
+                logger.atError().kv("secret", awsSecretResponse.getArn()).cause(e).log("Unable to decrypt secret, "
+                        + "skip loading in cache");
+                return;
+            }
+            String secretArn = decryptedResponse.arn();
+            cache.put(secretArn, decryptedResponse);
+            cache.put(secretArn + decryptedResponse.versionId(), decryptedResponse);
+            // load all labels attached with this version of secret
+            for (String label : decryptedResponse.versionStages()) {
+                cache.put(secretArn + label, decryptedResponse);
+            }
         }
     }
 
```

Comment:
We throw `GetSecretException` exception when customer [retrieves](https://github.com/aws-greengrass/aws-greengrass-secret-manager/blob/7ac16d4efdc13eef0e1f5f6bfe75c5bfec1efbf0/src/main/java/com/aws/greengrass/secretmanager/SecretManager.java#L260) a secret that is not configured as the secret is not available in the cache(here because it is not configured). 

If a secret is not available in the cache, we used to directly throw an exception stating that it is not found. With the [recent changes](https://github.com/aws-greengrass/aws-greengrass-secret-manager/blob/7ac16d4efdc13eef0e1f5f6bfe75c5bfec1efbf0/src/main/java/com/aws/greengrass/secretmanager/SecretManager.java#L251-L260), we try to download if it is configured and then return that secret. If it is not configured, we log why we couldn't download it and  then proceed by throwing an exception.

Reply:
Oh I see... so we validate if it's configured at both the refresh step as well as the return from cache step. There's no problem with the eventual behavior there but why not only check it once before doing either step? I.e. validate before acting

Reply:
Getting secret from cache doesn't validate if the secret is configured or not - we write to cache only if it is configured in the first place. Only while downloading the secret, we validate if it is configured because downloading secret can be triggered by customer's IPC request as well.